keps/sig-api-machinery/4153-declarative-validation/README.md

```diff
@@ -0,0 +1,1793 @@
+  will remain "server side only" validation rules.
+- It is not a goal to require that _all_ validation rules be written in CEL. We
+  will continue to support hand written validation rules indefinitely for use
+  cases that are a poor fit using the existing validation mechanism (or
+  something very similar).
+
+
+## Proposal
+
+Declarative validation will be performed against versioned APIs. This differs from the hand
+written validation, which is evaluated against internal types.
+
+Go IDL tags will be added to support the following declarative validation rules:
+
+| Type of valiation      | Go IDL tag                                                       | OpenAPI validation field                                                           | Reason Type      |
+| ---------------------- | ---------------------------------------------------------------- | ---------------------------------------------------------------------------------- | ---------------- |
+| string format          | `+format={format name}`                                          | `format`                                                                           | Invalid          |
+| size limits            | `+min{Length,Properties,Items}`, `+max{Length,Properties,Items}` | `min{Length,Properties,Items}`, `max{Length,Properties,Items}`                     | TooMany, TooLong |
+| numeric limits         | `+minimum`, `+maximum`, `+exclusiveMinimum`, `+exclusiveMaximum` | `minimum`, `maximum`, `exclusiveMinimum`, `exclusiveMaximum`                       | Invalid          |
+| required fields        | `+optional` (exists today)                                       | `required`                                                                         | Required         |
+| enum values            | `+enum` (exists today)                                           | `enum`                                                                             | NotSupported     |
+| uniqueness             | `listType=set` (sets and map keys)                               | `x-kubernetes-list-type`                                                           | Invalid          |
+| regex matches          | `+pattern`                                                       | `pattern`                                                                          | Invalid          |
+| cross field validation | `validationRule="{CEL expression}"`                              | `x-kubernetes-validations`                                                         | user specified   |
+| [transition rules][1]  | `validationRule="{CEL expression using oldSelf}"`                | `x-kubernetes-validations`                                                         | user specified   |
+| special case: name format | `+nameFormat={format name}`                                   | `x-kubernetes-validations` (see "format" section below for details)                | Invalid          |
```

Comment:
What about immutable (no updates) fields?

Reply:
What about mutually-exclusive fields (sometimes == Forbidden, sometimes Invalid) ?

Reply:
What about context-dependent fields (X may only be used if Y is a specific value) ?  Usually "Forbidden" but we could reconsider

Reply:
> What about immutable (no updates) fields?

This falls under `transition rules`

> What about mutually-exclusive fields (sometimes == Forbidden, sometimes Invalid) ?

This falls under `cross field validation`. The reason code (Forbidden/Invalid/etc) can also be customized using the `reason` field within `x-kubernetes-validations`

> What about context-dependent fields (X may only be used if Y is a specific value) ? Usually "Forbidden" but we could reconsider

Also `cross field validation`

Reply:
>> What about immutable (no updates) fields?
>
> This falls under transition rules

Sure, but do we force people to write `+validationRule="self == oldSelf",message="field is immutable"` or can they say `+immutable` and we generate the code and a consistent error on their behalf?

>> What about mutually-exclusive fields (sometimes == Forbidden, sometimes Invalid) ?
>
> This falls under cross field validation. 

Again, my question is really: If this is common (I don't have numbers but I know it's not zero), should we up-level this into `+mutually-exclusive=otherField` or something?

Reply:
I think that makes sense, especially for `+immutable`. We can desugar it into equivalent CEL on the code generator. Added a note that we will add IDL tags during development for idioms we identify.